### PR TITLE
Ep 1127 add deeplink logic to payment flow

### PR DESCRIPTION
**Pull request by marcghelas**

### CODE REVIEW

File: judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt [hunk 1]
```diff
@@ -61,6 +63,8 @@ class DemoFeatureListActivity : AppCompatActivity() {
 
         PreferenceManager.setDefaultValues(this, R.xml.root_preferences, false)
         sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
+
+        checkForDeepLink()
```
Comment (stpopa) [inline judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt:67]: let's rename this function to something like `deepLinkIfNeeded()`

File: judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt [hunk 2]
```diff
@@ -97,6 +101,28 @@ class DemoFeatureListActivity : AppCompatActivity() {
...
 
+    override fun onNewIntent(intent: Intent?) {
+        checkForDeepLink(intent)
+        super.onNewIntent(intent)
+    }
+
+    private fun checkForDeepLink(intent: Intent? = this.intent) {
+        val uri = intent?.data
+        val newIntent = Intent(this, JudoActivity::class.java)
+        if (uri != null) {
```
Comment (stpopa) [inline judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt:112]: there's no reason to create the `newIntent` if `uri == null`

Comment (stpopa) [reply judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt:112]: and we can just `intent?.data?.let { /* deep link stuff */}`

File: judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt [hunk 3]
```diff
@@ -97,6 +101,28 @@ class DemoFeatureListActivity : AppCompatActivity() {
...
+            val lastUsedPaymentWidget = sharedPreferences.getString(
```
Comment (stpopa) [inline judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt:113]:
```
val lastUsedPaymentWidget = sharedPreferences
.getString(LAST_USED_WIDGET_TYPE_LEY, null) ?: PaymentWidgetType.PAYMENT_METHODS.name
```

File: judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt [hunk 4]
```diff
@@ -97,6 +101,28 @@ class DemoFeatureListActivity : AppCompatActivity() {
...
+                LAST_USED_WIDGET_TYPE_LEY,
+                PaymentWidgetType.PAYMENT_METHODS.name
+            ) ?: PaymentWidgetType.PAYMENT_METHODS.name
+            val newPbbaConfig = pbbaConfiguration.newBuilder().setDeepLinkURL(uri).build()
```
Comment (stpopa) [inline judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt:117]: please let's get rid of this stuff: `newBuilder()`

File: judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt [hunk 5]
```diff
@@ -97,6 +101,28 @@ class DemoFeatureListActivity : AppCompatActivity() {
...
+            val judo = getJudo(PaymentWidgetType.valueOf(lastUsedPaymentWidget)).newBuilder()
```
Comment (stpopa) [inline judokit-android-examples/src/main/java/com/judokit/android/examples/feature/DemoFeatureListActivity.kt:118]: we don't this `newBuilder()`

File: judokit-android/src/main/java/com/judokit/android/Judo.kt [hunk 6]
```diff
@@ -177,6 +177,22 @@ class Judo internal constructor(
         }
     }
 
+    fun newBuilder() = Builder(paymentWidgetType)
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/Judo.kt:180]: we don't need this

File: judokit-android/src/main/java/com/judokit/android/model/PBBAConfiguration.kt [hunk 7]
```diff
@@ -1,32 +1,58 @@
...
 
         fun setEmailAddress(emailAddress: String?) = apply { this.emailAddress = emailAddress }
+
+        fun setDeepLinkURL(deepLinkURL: Uri?) = apply { this.deepLinkURL = deepLinkURL }
+
+        fun setDeepLinkScheme(deepLinkScheme: String?) =
+            apply { this.deepLinkScheme = deepLinkScheme }
+
+        fun build(): PBBAConfiguration {
+            val myDeepLinkScheme = requireNotNull(deepLinkScheme)
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/model/PBBAConfiguration.kt:36]: what happens when merchant doesn't use PbBA as a payment method?

File: judokit-android/src/main/java/com/judokit/android/service/polling/PollingResult.kt [hunk 8]
```diff
@@ -6,11 +6,11 @@ sealed class PollingResult<out T> {
     object Delay : PollingResult<Nothing>()
     object Retry : PollingResult<Nothing>()
     object Processing : PollingResult<Nothing>()
-    data class Failure(
+    data class CallFailure(
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/service/polling/PollingResult.kt:9]: we'll need to refactor this stuff at some point

File: judokit-android/src/main/java/com/judokit/android/ui/paymentmethods/PaymentMethodsFragment.kt [hunk 9]
```diff
@@ -73,7 +75,13 @@ class PaymentMethodsFragment : Fragment() {
 
     override fun onActivityCreated(savedInstanceState: Bundle?) {
         super.onActivityCreated(savedInstanceState)
-
+        if (judo.pbbaConfiguration?.deepLinkURL != null) {
+            findNavController().navigate(
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/ui/paymentmethods/PaymentMethodsFragment.kt:79]: this is duplicated on line 120, please extract it to a method and reuse it

File: judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusFragment.kt [hunk 10]
```diff
@@ -60,13 +61,25 @@ class PollingStatusFragment : DialogFragment(), PBBAPopupCallback {
...
+            application,
+            judo,
+            paymentWidgetType
+        )
         viewModel = ViewModelProvider(this, factory).get(PollingStatusViewModel::class.java)
 
-        viewModel.send(PollingAction.PayWithPayByBank)
+        viewModel.send(PollingAction.Initialise(sharedViewModel.error.details.isEmpty()))
+
+        viewModel.viewObserver.observe(viewLifecycleOwner, Observer {
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusFragment.kt:80]: please use a more informative name `viewObserver` says nothing

File: judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusViewModel.kt [hunk 11]
```diff
@@ -12,14 +13,18 @@ import com.judokit.android.api.model.request.BankSaleRequest
...
 import com.judokit.android.api.model.response.BankSaleStatusResponse
 import com.judokit.android.api.model.response.JudoApiCallResult
+import com.judokit.android.model.PaymentWidgetType
 import com.judokit.android.service.polling.PollingResult
 import com.judokit.android.service.polling.PollingService
 import com.judokit.android.toMap
+import com.judokit.android.ui.paymentmethods.model.Event
 import kotlinx.coroutines.launch
 
+// TODO: Change to orderId
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusViewModel.kt:23]: why not to change it now?

File: judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusViewModel.kt [hunk 12]
```diff
@@ -43,21 +55,50 @@ class PollingStatusViewModel(
...
+    private fun retry() {
+        viewModelScope.launch {
+            pollingService.retry()
+        }
+    }
+
+    private fun handleDeepLinkCallback(url: Uri) {
+        val orderId = url.getQueryParameter(ORDER_ID)
+        if (orderId != null) {
+            viewObserver.postValue(Event())
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusViewModel.kt:98]: why we need this at all?

File: judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusFragment.kt [hunk 13]
```diff
@@ -60,13 +61,25 @@ class PollingStatusFragment : DialogFragment(), PBBAPopupCallback {
...
+            service,
+            pollingService,
+            application,
+            judo,
+            paymentWidgetType
+        )
         viewModel = ViewModelProvider(this, factory).get(PollingStatusViewModel::class.java)
 
-        viewModel.send(PollingAction.PayWithPayByBank)
+        viewModel.send(PollingAction.Initialise(sharedViewModel.error.details.isEmpty()))
```
Comment (stpopa) [inline judokit-android/src/main/java/com/judokit/android/ui/pollingstatus/PollingStatusFragment.kt:78]: this `sharedViewModel.error.details.isEmpty() ` is obscure, why `isEmpty == isDeepLinkCallback ` ?

File: judokit-android/src/main/java/com/judokit/android/service/polling/PollingResult.kt (hunk 8 above)
Comment (marcghelas) [reply judokit-android/src/main/java/com/judokit/android/service/polling/PollingResult.kt:9]: Will be addressed in the future

**Pull request merged**